contracts/RestrictedNORI.sol

```diff
@@ -0,0 +1,1089 @@
+   *
+   * @dev Follows the rules of hooks defined [here](
+   *  https://docs.openzeppelin.com/contracts/4.x/extending-contracts#rules_of_hooks)
+   * @dev See the ERC1155 specific version [here](
+   *  https://docs.openzeppelin.com/contracts/3.x/api/token/erc1155#ERC1155-_beforeTokenTransfer-address-address-address-uint256---uint256---bytes-)
+   *
+   * ##### Requirements:
+   *
+   * - the contract must not be paused
+   * - One of the following must be true:
+   *    - the operation is a mint (which should ONLY occur when BridgedPolygonNORI is being wrapped via `_depositFor`)
+   *    - the operation is a burn, which only happens during revocation and withdrawal:
+   *      - if the operation is a revocation, that permission is enforced by the TOKEN_REVOKER_ROLE
+   *      - if the operation is a withdrawal the burn amount must be <= the sender's claimable balance
+   *    - the operation is a transfer and _all_ of the following must be true:
+   *      - the operator is operating on their own balance (enforced in the inherited contract)
+   *      - the operator has sufficient balance to transfer (enforced in the inherited contract)
+   */
+  function _beforeTokenTransfer(
+    address operator,
+    address from,
+    address to,
+    uint256[] memory ids,
+    uint256[] memory amounts,
+    bytes memory data
+  )
+    internal
+    override(ERC1155PresetMinterPauserUpgradeable, ERC1155SupplyUpgradeable)
+  {
+    bool isMinting = from == address(0);
+    bool isBurning = to == address(0);
+
+    bool isWithdrawing = isBurning && from == operator;
+    bool isTransferring = !isMinting && !isBurning;
+
+    if (isBurning) {
+      for (uint256 i = 0; i < ids.length; i++) {
+        Schedule storage schedule = _scheduleIdToScheduleStruct[ids[i]];
+        schedule.releasedAmountFloor = _releasedBalanceOfSingleSchedule(ids[i]);
+      }
+    }
+    if (isWithdrawing) {
+      for (uint256 i = 0; i < ids.length; i++) {
+        if (amounts[i] > claimableBalanceForScheduleForAccount(ids[i], from)) {
+          revert InsufficientClaimableBalance({
+            account: from,
+            scheduleId: ids[i]
+          });
+        }
+      }
+    }
+    return super._beforeTokenTransfer(operator, from, to, ids, amounts, data);
+  }
+
+  /**
+   * Linearly released balance for a single schedule at the current block timestamp, ignoring any
+   * released amount floor that has been set for the schedule.
+   */
+  function _linearReleaseAmountAvailable(uint256 scheduleId)
+    internal
+    view
+    returns (uint256)
+  {
+    Schedule storage schedule = _scheduleIdToScheduleStruct[scheduleId];
+    if (block.timestamp >= schedule.endTime) {
+      return totalSupply(scheduleId);
+    }
+    uint256 rampTotalTime = schedule.endTime - schedule.startTime;
+    return
+      block.timestamp < schedule.startTime
+        ? 0
+        : (totalSupply(scheduleId) * (block.timestamp - schedule.startTime)) /
+          rampTotalTime;
```

Comment:
this is a little hard to read, perhaps this could be assigned to a variable and then rewritten to only include a single return statement